examples/fluids/src/velocity_gradient_projection.c

```diff
@@ -0,0 +1,173 @@
+  CeedOperatorSetField(op_rhs_assemble, "qdata", ceed_data->elem_restr_qd_i, CEED_BASIS_COLLOCATED, ceed_data->q_data);
+  CeedOperatorSetField(op_rhs_assemble, "x", ceed_data->elem_restr_x, ceed_data->basis_x, ceed_data->x_coord);
+  CeedOperatorSetField(op_rhs_assemble, "velocity gradient", elem_restr_grad_velo, basis_grad_velo, CEED_VECTOR_ACTIVE);
+
+  CeedElemRestrictionCreateVector(ceed_data->elem_restr_q, &q_ceed, NULL);
+  CeedElemRestrictionCreateVector(elem_restr_grad_velo, &rhs_ceed, NULL);
+
+  PetscCall(
+      OperatorApplyContextCreate(user->dm, grad_velo_proj->dm, ceed, op_rhs_assemble, q_ceed, rhs_ceed, NULL, NULL, &grad_velo_proj->l2_rhs_ctx));
+
+  // -- Build Mass operator
+  PetscCall(CreateMassQFunction(ceed, grad_velo_proj->num_comp, q_data_size, &qf_mass));
+  CeedOperatorCreate(ceed, qf_mass, NULL, NULL, &op_mass);
+  CeedOperatorSetField(op_mass, "u", elem_restr_grad_velo, basis_grad_velo, CEED_VECTOR_ACTIVE);
+  CeedOperatorSetField(op_mass, "qdata", ceed_data->elem_restr_qd_i, CEED_BASIS_COLLOCATED, ceed_data->q_data);
+  CeedOperatorSetField(op_mass, "v", elem_restr_grad_velo, basis_grad_velo, CEED_VECTOR_ACTIVE);
+
+  {  // -- Setup KSP for L^2 projection with lumped mass operator
+    PetscInt l_size, g_size;
+    Mat      mat_mass;
+    VecType  vec_type;
+    Vec      M_inv;
+    MPI_Comm comm = PetscObjectComm((PetscObject)grad_velo_proj->dm);
+
+    PetscCall(DMGetGlobalVector(grad_velo_proj->dm, &M_inv));
+    PetscCall(VecGetLocalSize(M_inv, &l_size));
+    PetscCall(VecGetSize(M_inv, &g_size));
+    PetscCall(VecGetType(M_inv, &vec_type));
+    PetscCall(DMRestoreGlobalVector(grad_velo_proj->dm, &M_inv));
+
+    CeedElemRestrictionCreateVector(elem_restr_grad_velo, &mass_output, NULL);
+    PetscCall(OperatorApplyContextCreate(grad_velo_proj->dm, grad_velo_proj->dm, ceed, op_mass, rhs_ceed, mass_output, NULL, NULL, &mass_matop_ctx));
+    CeedVectorDestroy(&mass_output);
+
+    PetscCall(MatCreateShell(comm, l_size, l_size, g_size, g_size, mass_matop_ctx, &mat_mass));
+    PetscCall(MatShellSetContextDestroy(mat_mass, (PetscErrorCode(*)(void *))OperatorApplyContextDestroy));
+    PetscCall(MatShellSetOperation(mat_mass, MATOP_MULT, (void (*)(void))MatMult_Ceed));
+    PetscCall(MatShellSetOperation(mat_mass, MATOP_GET_DIAGONAL, (void (*)(void))MatGetDiag_Ceed));
+    PetscCall(MatShellSetVecType(mat_mass, vec_type));
+
+    PetscCall(KSPCreate(comm, &grad_velo_proj->ksp));
+    PetscCall(KSPSetOptionsPrefix(grad_velo_proj->ksp, "velocity_gradient_projection_"));
+    {
+      PC pc;
+      PetscCall(KSPGetPC(grad_velo_proj->ksp, &pc));
+      PetscCall(PCSetType(pc, PCJACOBI));
+      PetscCall(PCJacobiSetType(pc, PC_JACOBI_ROWSUM));
+      PetscCall(KSPSetType(grad_velo_proj->ksp, KSPPREONLY));
+      // TODO Not sure if the option below are necessary
+      PetscCall(KSPSetConvergenceTest(grad_velo_proj->ksp, KSPConvergedSkip, NULL, NULL));
```

Comment:
As comment says, not sure if this is necessary or not. I notice that if I monitor the convergence, it does report residuals before and after preconditioning application, which makes me worried that it's actually calculating the residual here. 

Any ideas how to skip the residual calculation? Or if it's actually happening at all?

Reply:
> Also note that ROWSUM is fragile since it won't work for high order tets.

Right. There isn't another lumped-mass matrix option right now, is there? I'm pretty sure "special lumping" would get around that issue if/when we run into it.

Reply:
Yeah, we'll need to revisit more than just here too if we run into lumped mass matrix issues.

We could also do a collocated quadrature approach and then just use `CeedOperatorLinearAssembleDiagonal`, right? Then we'd could just hand off that vector to the KSP as the preconditioning matrix.

Reply:
Maybe we should fork this conversation off into a GH issue? Just to have a central location to discuss, when we do get to it.

Reply:
Yeah, can fork. Collocated quadrature doesn't fix the issue with rowsum lumping.